### ISSUE TITLE
transfer learning with existing class

### ISSUE THREAD

**Issue by iFamZ** [matterport/Mask_RCNN#2364]
I am trying to implement transfer learning with Mask_RCNN on a new dataset. The dataset is labeled with a class that already exist in coco (boat). After following a few guides, I was able to use a new dataset for the transfer learning. However, when I try to use my new weights for inference, I am pretty much given garbage. 

I ran the training ran for 30 epochs with 100 steps per epoch. I set the minimum confidence to 0.9 and enabled the image padding. However, for the number of classes, I used 2 - the guides I used had it set to 2 (should this be 1 for just the background? since the boat class is already a class in coco?). I also had to remove certain layers in my network, specifically mrcnn_class_logits, mrcnn_bbox_fc, mrcnn_bbox, and mrcnn_mask. I had to remove these layers because the shapes did not match and I would be thrown an error.

Is there anything that is glaringly wrong? Can you guys point me to a direction where the problem might lie? If any of you have a comprehensive guide for transfer learning for a class that already exists, I would greatly appreciate a link to that.

I appreciate any help I can receive and thank you for your time as well.

Thanks!

**Comment by suchiz** [comment]
Hello, 
30 epochs with only 100 step per epoch is very low. It should explain why you get garbage.
Indeed, you have to set the number of classes to 2: background + boat
Transfer learning in Mask-RCNN consist on keeping the COCO backbone (features extractor) + COCO RPN (object detection). So you have to train the 'heads' only. But as I said, 30 epochs with only 100 step per epoch is very low.

**Comment by iFamZ** [comment]
I appreciate the comment.

I understand that the number of epochs and the number of training samples are very low, shouldn't that still improve my performance? Shouldn't the algorithm know more about these synthetic images from the 10 additional images than it did when it didn't see any synthetic images at all? In fact, even when I try to use a training sample as a testing input, I still get bad results. The results are actually better with the out of the box weights, than the weights after my training. I am really perplexed by this - am I missing something?

For instance, the out of the box will correctly label the image, which has only one class. Whereas, the transfer learned weights would label the image with other classes, which do not exist, and miss the class which does.

I will be obtaining hardware to be able to process more training samples with more epochs. I can update the status then, but unfortunately, currently I only have access to outdated hardware.

Again, I appreciate your reply! Thanks!

**Comment by suchiz** [comment]
Hmm, yeah you're right. And in transfer learning, 30 epoch with 100 should be enough to see something. Hence, something is wrong with your training. Checkout if the labels and masks are well loaded. If it is, check if the data augmentation isn't too strong.

**Comment by burhr2** [comment]
Maybe try to use the inspect_data in samples/balloon folder to check if your dataset is loaded properly. If it is loaded properly with masks and bbox then you know its the setup that maybe the problem. Because even with small images, with transfer learning like the tutorial example should give some reasonable results.
